### ISSUE TITLE
[client]: client abstraction proposal

### ISSUE THREAD

**Issue by niklasad1** [paritytech/jsonrpsee#288]
After #269 the transports are abstract i.e. doesn't know anything related JSON-RPC details it just sends raw messages which makes it possible to bring provide two abstract clients:

- Basic client - a client that doesn't do subscriptions and don't need a background thread (similar to the current HTTP client)
- Subscription client/async client - a client that needs a background task to handle requests, subscriptions and so on (similar to the current WebSocket client)

Thus, if we could provide two different transport traits for that it should be really easy to add new transports without duplicating the logic.

Example:

```
 /// Transport that sends out a message and waits for its response. 
 pub trait BlockingTransport {
    type Error;

    // TODO: handle `&mut self`
    fn send(&self, message: String) -> Result<String, Self::Error>;
 }

 /// Transport that sends messages.
 pub trait AsyncTransportSender {
    type Error;

    // TODO: handle `&mut self`
    fn send(&self, message: String) -> Result<(), Self::Error>;
}

 /// Transport that reads messages.
 pub trait AsyncTransportReceiver {
    type Error;

    // TODO: handle `&mut self`
    fn recv(&self, message: String) -> Result<String, Self::Error>;
 }
```

**Comment by dvdplm** [comment]
@niklasad1 is this still relevant for v0.4 you think?

**Comment by niklasad1** [comment]
> @niklasad1 is this still relevant for v0.4 you think?

yes

**Comment by niklasad1** [comment]
Closed by #580.
